test/performance/metrics/runtime.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"log"
+	"time"
+
+	"k8s.io/apimachinery/pkg/labels"
+	deploymentinformer "knative.dev/pkg/injection/informers/kubeinformers/appsv1/deployment"
+	netv1alpha1 "knative.dev/serving/pkg/apis/networking/v1alpha1"
+	sksinformer "knative.dev/serving/pkg/client/injection/informers/networking/v1alpha1/serverlessservice"
+
+	// Mysteriously required to support GCP auth (required by k8s libs).
+	// Apparently just importing it is enough. @_@ side effects @_@. https://github.com/kubernetes/client-go/issues/242
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+)
+
+// DeploymentStatus is a struct that wraps the status of a deployment.
+type DeploymentStatus struct {
+	DesiredReplicas int32
+	ReadyReplicas   int32
+	// Time is the time when the status is fetched
+	Time time.Time
+}
+
+// FetchDeploymentStatus creates a channel that can return the up-to-date DeploymentStatus periodically.
+func FetchDeploymentStatus(
+	ctx context.Context, namespace string, selector labels.Selector,
+	duration time.Duration,
+) <-chan DeploymentStatus {
+	dl := deploymentinformer.Get(ctx).Lister()
```

Comment:
If we have informers at our disposal, why work with tickers at all? Can we not be informed if something changed and make this more event-driven rather than poll-based?

Reply:
event-driven sounds a better approach to me as well, as we will not miss any changing points. @mattmoor WDYT?

Reply:
I thought about this a bit, and initially I was going to say: "drop the duration arg, and we can change the backing implementation." but...

I think that to some extent it depends on what we want these metrics for.  By reporting at a fixed interval, things like average pod count are meaningful, but with a variable resolution they are not.  The lines will also render improperly because it will directly connect distant points vs. plateau and fall.

So I'd leave this implementation alone (I think it's worth having), but I'm open to discussing alternatives.